include/swift/Parse/Parser.h

```diff
@@ -1325,6 +1326,49 @@ class Parser {
   /// Get the location for a type error.
   SourceLoc getTypeErrorLoc() const;
 
+  /// Callback function used for creating a C++ AST from the syntax node at the given source location.
+  ///
+  /// The arguments to this callback are the source file to pass into ASTGen (the exported source file)
+  /// and the source location pointer to pass into ASTGen (to find the syntax node).
+  ///
+  /// The callback returns the new AST node and the ending location of the syntax node. If the AST node
+  /// is NULL, something went wrong.
+  template<typename T>
+  using ASTFromSyntaxTreeCallback = std::pair<T*, const void *>(
+      void *sourceFile, const void *sourceLoc
```

Comment:
We really need to figure out a better way to pass things around than using `void *`.

Reply:
Yes, this is pretty dreadful. Typedefs would improve on things a little bit. @bnbarham keeps threatening to add them

Reply:
https://github.com/apple/swift/pull/66044 for the common types (though this would be the first `SourceFile` I think?). Decls/exprs/typerefs/etc are a little more difficult since the visitor just returns a `ASTNode` for each `visit` at the moment. Even disregarding that I'm not sure we want to bridge every single AST node type that we need.

But it would be nice to do something to avoid:
  - https://github.com/apple/swift/pull/66033/commits/682ac98b253e1ef18ffc44e6f6c767a02ca79125 - forgot `.rawValue`
  - https://github.com/apple/swift/pull/66044/files#diff-05684b1c93b20e9444846f61a3575d650a447713a129b477ecf5a518828d17e9L187 - `name` was *not* a `BridgedIdentifier` but rather a `UnresolvedDeclRefExpr`. Finding that involves looking at the visit for `VariableDeclSyntax` -> visits its `pattern` -> visits `IdentifierPatternSyntax` -> calls `SwiftIdentifierExpr_create` -> creates a `UnresolvedDeclRefExpr`.